### ISSUE TITLE
Is there a function which performs the opposite of tf.extract_image_patches ? 

### ISSUE THREAD

**Issue by VigneshSrinivasan10** [tensorflow/tensorflow#6743]
Hello, 

I was looking for a function in the latest tf version, which performs the opposite of tf.extract_image_patches. If I did not miss it, it would be great to have this feature in tensorflow. 
It would help me for the task I am doing.  

Thanks in advance,
Vignesh

**Comment by bodokaiser** [comment]
+1

**Comment by bodokaiser** [comment]
@VigneshSrinivasan10 I found it hard to work with `tf.extract_image_patches` as it does a lot in the background. If your patches are just non-overlapping blocks then I found it easier to use `tf.reshape` and `tf.transpose`:

```python
def image_to_patches(image, image_height, image_width, patch_height, patch_width):
    height = math.ceil(image_height/patch_height)*patch_height
    width = math.ceil(image_width/patch_width)*patch_width

    image_resized = tf.squeeze(tf.image.resize_image_with_crop_or_pad(image, height, width))
    image_reshaped = tf.reshape(image_resized, [height // patch_height, patch_height, -1, patch_width])
    image_transposed = tf.transpose(image_reshaped, [0, 2, 1, 3])
    return tf.reshape(image_transposed, [-1, patch_height, patch_width, 1]
```
See [stackoverflow](http://stackoverflow.com/questions/41564321/split-image-tensor-into-small-patches) for more.

**Comment by VigneshSrinivasan10** [comment]
@bodokaiser, `tf.extract_image_patches` in my case can also work with overlap. I am looking to add the values in the places where they overlap.

**Comment by bodokaiser** [comment]
@VigneshSrinivasan10 in this case I think its best to put all `patches` belonging to the same `image` in one `batch` reshaping it to a 2d-tensor and then using `tf.nn.conv2d` to handle the overlap.

As this is not trivial to implement and something which I think occurs quite often it would really be great to have this supported by tensorflow.

**Comment by VigneshSrinivasan10** [comment]
@bodokaiser, I totally agree. It will be great to have this supported by tensorflow

**Comment by jart** [comment]
@gpapan two users are requesting this feature. Since you introduced the ExtractImagePatches op, I figured you might be a good person to weigh in on the merits.

**Comment by gpapan** [comment]
The opposite of `tf.extract_image_patches` is its gradient or transpose operation.
There is an implementation of that [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/array_grad.py#L575).
Also see: #2921.

**Comment by bodokaiser** [comment]
What do I need to do to create API docs for this?

**Comment by jart** [comment]
I'm going to mark this as resolved, per @gpapan's response. If you feel this should still be actionable on the part of TensorFlow, let me know and I'll re-open this issue.

**Comment by bodokaiser** [comment]
@jart @gpapan it is still not very clear to me how you use that function. I also do not find anything in the API docs about it.

**Comment by jart** [comment]
@bodokaiser we recommend [StackOverflow](http://stackoverflow.com/questions/tagged/tensorflow) for community-driven support.

**Comment by bodokaiser** [comment]
I still see this as a documentation issue - as mentioned earlier I could fix that with a PR if I know where to start.

**Comment by jart** [comment]
In that case, please open up a new issue requesting more in-depth documentation for `tf.extract_image_patches`. I'll triage it right away and make sure the appropriate people see it.

**Comment by bodokaiser** [comment]
Thank you @jart! you find the corresponding request at #6847.

**Comment by shaolinkhoa** [comment]
Hi all,
I found this topic and #6847

**Comment by LorSong** [comment]
I've came up with naive solution
Sice I wanted mean of overlapping pixels, we can add all pixels in proper location and divide by the number of addition (overlaps). 

*   Create 2 arrays of zeros of original image size (black image)
*   One array will be image where we adding our patches, second will be the "mask", counting how many patches that pixel recieved.
*   Divide image by mask

Full code with slicing first, then reversing
Slicing

`
image, n_images_v, n_images_h, kernel_size = #your image, your choice
steps = [(image.shape[i] - kernel_size[i]) // (n_images_v - 1) for i in range(2)]

# Making slices
slices = tf.image.extract_patches(tf.expand_dims(image, axis=0), # adding batch axis
              sizes=[1,kernel_size[0], kernel_size[1],1], 
              strides=[1, steps[0], steps[1], 1], 
              rates=[1,1,1,1], 
              padding="VALID")
print("Shape of resulted slices: ", slices.shape.as_list())
# Reshaping into images
sliced_images = tf.reshape(slices, (-1, kernel_size[0], kernel_size[1]))`

Revercing
`# Initialize image
recovered = tf.zeros(image.shape[0:2])
mask = tf.zeros(image.shape[0:2], dtype=tf.int8)
# Kernel for counting overlaps
kernel_ones = tf.ones(kernel_size,dtype=tf.int8)
for j in range(n_images_v):
  for i in range(n_images_h):
    # Make indices from meshgrid
    indexes = tf.meshgrid(tf.range(steps[0] * j, # row start
                                   kernel_size[0] + steps[0] * j), # row end
                          tf.range(steps[1] * i, # col_start
                                   kernel_size[1] + steps[1] * i), indexing='ij') # col_end
    indexes = tf.stack(indexes, axis=-1)
    # Add sliced image to recovered image indice
    recovered = tf.tensor_scatter_nd_add(recovered, indexes, sliced_images[i+j*3])
    # Update mask
    mask = tf.tensor_scatter_nd_add(mask, indexes, kernel_ones)

recovered = recovered / tf.cast(mask, tf.float32)

plot_gray_images([recovered.numpy()])`

**Comment by tristan-deep** [comment]
Building on @LorSong's answer, I made a function that is able to perform the reverse of  `tf.image.extract_patches` but on 4D tensors (with batch and channel axis). Overlapping patches are supported, and can either replace or average between overlapping parts.

```python
def patches_to_images_tf(
    patches: np.ndarray, image_shape: tuple,
    overlap: float = 0.5, stitch_type='average', indices=None) -> np.ndarray:
    """Reconstructs images from patches.

    Args:
        patches (ndarray): Array with batch of patches to convert to batch of images.
            [batch_size, #patch_y, #patch_x, patch_size_y, patch_size_x, n_channels]
        image_shape (Tuple): Shape of output image. (y, x, n_channels) or (y, x)
        overlap (float, optional): Overlap factor between patches. Defaults to 0.5.
        stitch_type (str, optional): Type of stitching to use. Defaults to 'average'.
            Options: 'average', 'replace'.
        indices (ndarray, optional): Indices of patches in image. Defaults to None.
            If provided, indices are used to stitch patches together and not recomputed
            to save time. Has same shape as patches shape but with added index axis (last).
    Returns:
        images (ndarray): Reconstructed batch of images from batch of patches.

    """
    assert len(image_shape) == 3, 'image_shape should have 3 dimensions, namely: ' \
        '(#image_y, #image_x, (n_channels))'
    assert len(patches.shape) == 6 , 'patches should have 6 dimensions, namely: ' \
        '[batch, #patch_y, #patch_x, patch_size_y, patch_size_x, n_channels]'
    assert overlap >= 0 and overlap < 1, 'overlap should be between 0 and 1'
    assert stitch_type in ['average', 'replace'], 'stitch_type should be either ' \
        '"average" or "replace"'

    batch_size, n_patches_y, n_patches_x, *patch_shape = patches.shape
    n_channels = image_shape[-1]
    dtype = patches.dtype

    patches_fit, new_image_shape = check_patches_fit(image_shape[:2], patch_shape[:2], overlap)
    if not patches_fit:
        image_shape = (*new_image_shape, image_shape[-1])

    assert len(patch_shape) == len(image_shape)

    # Kernel for counting overlaps
    if stitch_type == 'average':
        kernel_ones = tf.ones((batch_size, n_patches_y, n_patches_x, *patch_shape), dtype=tf.int32)
        mask = tf.zeros((batch_size, *image_shape), dtype=tf.int32)

    if indices is None:
        if overlap:
            nonoverlap = np.array([1 - overlap, 1 - overlap, 1 / patch_shape[-1]])
            stride = (np.array(patch_shape) * nonoverlap).astype(int)
        else:
            stride = (np.array(patch_shape)).astype(int)

        channel_idx = tf.reshape(tf.range(n_channels), (1, 1, 1, 1, 1, n_channels, 1))
        channel_idx = (tf.ones((batch_size, n_patches_y, n_patches_x, *patch_shape, 1), dtype=tf.int32) * channel_idx)

        batch_idx = tf.reshape(tf.range(batch_size), (batch_size, 1, 1, 1, 1, 1, 1))
        batch_idx = (tf.ones((batch_size, n_patches_y, n_patches_x, *patch_shape, 1), dtype=tf.int32) * batch_idx)

        # TODO: create indices without looping possibly
        indices = []
        for j in range(n_patches_y):
            for i in range(n_patches_x):
                # Make indices from meshgrid
                _indices = tf.meshgrid(
                    tf.range(stride[0] * j, # row start
                            patch_shape[0] + stride[0] * j), # row end
                    tf.range(stride[1] * i, # col_start
                            patch_shape[1] + stride[1] * i), indexing='ij') # col_end

                _indices = tf.stack(_indices, axis=-1)
                indices.append(_indices)

        indices = tf.reshape(tf.stack(indices, axis=0), (n_patches_y, n_patches_x, *patch_shape[:2], 2))

        indices = tf.repeat(indices[tf.newaxis, ...], batch_size, axis=0)
        indices = tf.repeat(indices[..., tf.newaxis, :], n_channels, axis=-2)

        indices = tf.concat([batch_idx, indices, channel_idx], axis=-1)

    # create output image tensor
    images = tf.zeros([batch_size, *image_shape], dtype=dtype)

    # Add sliced image to recovered image indices
    if stitch_type == 'replace':
        images = tf.tensor_scatter_nd_update(images, indices, patches)
    else:
        images = tf.tensor_scatter_nd_add(images, indices, patches)
        mask = tf.tensor_scatter_nd_add(mask, indices, kernel_ones)
        images = tf.cast(images, tf.float32) / tf.cast(mask, tf.float32)

    return images, indices
```

**Comment by gvalenzuelarg** [comment]
Thank you, @tristan-deep! This is just what I was looking for.
